### PR TITLE
Added BitmapBrush

### PR DESCRIPTION
**Pull request by beklemeto**
This pull request adds a BitmapBrush. Tested and used successfully for months with v1.4.1. I tried to port the changes to the master branch. Sorry for the untidy pull-request, as it's my first here on github.

### CODE REVIEW

File: src/D2DLibExport/D2DEnums.cs [hunk 1]
```diff
@@ -345,7 +345,28 @@ public enum D2DBitmapInterpolationMode
...
 		Linear = 1,
 	}
-
+ 
+	public enum D2DExtendMode
+	{
+		/// 
+		/// Extend the edges of the source out by clamping sample points outside the source
+		/// to the edges.
+		///
```
Comment (drewnoakes) [inline src/D2DLibExport/D2DEnums.cs:354]: Using `<summary>` will allow these docs to appear in IDE tooling etc.

```suggestion
		/// <summary>
		/// Extend the edges of the source out by clamping sample points outside the source
		/// to the edges.
		/// </summary>
```

Ditto below.

File: src/d2dlib/Context.h [hunk 2]
```diff
@@ -129,7 +131,7 @@ inline void SafeRelease(Interface **ppInterfaceToRelease)
 }
 
 #define RetrieveContext(ctx) D2DContext* context = reinterpret_cast<D2DContext*>(ctx)
-#define RetrieveD2DBitmap(ctx, handle) ID2D1Bitmap* ctx = reinterpret_cast<ID2D1Bitmap*>(handle)
+#define RetrieveD2DBitmap(ctx) ID2D1Bitmap* d2dbitmap = reinterpret_cast<ID2D1Bitmap*>(d2dbitmapHandle)
```
Comment (drewnoakes) [inline src/d2dlib/Context.h:134]: Is this right? It seems to rely upon there being a `d2dbitmapHandle` value in scope, where the previous code (and the `#define` above) only rely on their "arguments".

File: src/D2DLibExport/D2DBitmapBrush.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace unvell.D2DLib
+{
+  public class D2DBitmapBrush : D2DBrush
+  {
+    public D2DBitmap Bitmap { get; private set; }
```
Comment (drewnoakes) [inline src/D2DLibExport/D2DBitmapBrush.cs:34]: I think this can be get-only.

```suggestion
    public D2DBitmap Bitmap { get; }
```

**Review by drewnoakes — COMMENTED** [review]
Looks great though @jingwood should sign off.

Would be amazing if you could add an example of using this to one of the demo apps.

**Comment by jingwood** [comment]
Hi @beklemeto! Thanks very much for your work on this.

Could you please address the points that @drewnoakes raised? I'll merge this PR once those issues are resolved.

Additionally, if you could provide an example program, that would be great. It will help us better understand how your feature works.

File: src/d2dlib/Context.h (hunk 2 above)
Comment (jingwood) [reply src/d2dlib/Context.h:134]: It seems that both arguments, `contextHandle` and `bitmapHandle`, should be used. However, `RetrieveD2DBitmap` is currently not being utilized anywhere. I plan to fix this in the future.

**Comment by jingwood** [comment]
I'm going to merge this and make a simple example code to test it.

**Comment by beklemeto** [comment]
> I'm going to merge this and make a simple example code to test it.

Thank you for dealing with it. I'm not able to react swiftly.

Maybe it would be useful to point my exact code excerpt on how I used it:
```
public void DrawPatternPolygon(D2DPoint[] points, Bitmap pattern)
{
	if (graphics != null)
	{
		using (var bmp = panel.Device.CreateBitmapGraphics(pattern.Width, pattern.Height))
		{
			using (var d2dbitmap = panel.Device.LoadBitmap(ImageToByte(pattern)))
			{
				bmp.BeginRender();
				bmp.DrawBitmap(d2dbitmap, new D2DRect(0, 0, pattern.Width, pattern.Height), 1f);
				bmp.EndRender();

				using (var patternBrush = panel.Device.CreateBitmapBrush(bmp.GetBitmap(), D2DExtendMode.Wrap, D2DExtendMode.Wrap, D2DBitmapInterpolationMode.Linear))
				{
					graphics.DrawPolygon(points, D2DColor.Black, 0, D2DDashStyle.Solid, patternBrush);
				}
			}
		}
	}
}
```

The code repeats a pattern horizontally and vertically (Wrap extend mode) as long as the polygon is fully filled. Please note the double Bitmap conversion through byte array and then drawing onto a bitmap graphics. This was the only way possible to convert **transparent** GDI Bitmap into a **transparent** D2DBitmap.

Excellen description on how to use the extend parameters of the brush can be found here:
https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src/Direct2D/direct2d-brushes-overview.md

**Pull request merged**